### PR TITLE
clients/horizonclient: fix multi-parameter url for claimable balance query

### PR DESCRIPTION
**Pull request by iateadonut**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [X] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [X] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [X] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [X] This PR adds tests for the most critical parts of the new functionality or fixes.
* [ ] I've updated any docs ([developer docs](https://developers.stellar.org/api/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [ ] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [ ] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### What

bug: https://github.com/stellar/go/issues/4247

Fix cbr.BuildURL() so that it can build a multi-parameter url if one of the url's is not ID.

### Why

Bug does not allow the proper claimable balance query to be made.

### Known limitations

N/A

### CODE REVIEW

**Review by daesu — APPROVED** [review]
lgtm

File: clients/horizonclient/claimable_balance_request.go [hunk 1]
```diff
@@ -16,19 +17,24 @@ func (cbr ClaimableBalanceRequest) BuildURL() (endpoint string, err error) {
 	// Only one filter parameter is allowed, and you can't mix an ID query and
 	// filters.
 	nParams := countParams(cbr.Asset, cbr.Claimant, cbr.Sponsor, cbr.ID)
-	if nParams > 1 {
+	if cbr.ID != "" && nParams > 1 {
```
Comment (Shaptic) [inline clients/horizonclient/claimable_balance_request.go:20]: I'm not sure this is necessary? `countParams` returns 1 if the `cbr.ID` is set.

File: clients/horizonclient/claimable_balance_request.go [hunk 2]
```diff
@@ -16,19 +17,24 @@ func (cbr ClaimableBalanceRequest) BuildURL() (endpoint string, err error) {
...
 	} else {
+		params := map[string]string{
+			"claimant": cbr.Claimant,
+			"sponsor":  cbr.Sponsor,
+			"asset":    cbr.Asset,
+			"cursor":   cbr.Cursor,
+		}
+		if (cbr.Limit) > 0 {
+			params["limit"] = strconv.FormatUint(uint64(cbr.Limit), 10)
+		}
```
Comment (Shaptic) [inline clients/horizonclient/claimable_balance_request.go:35]: Let's let the backend handle the defaults

File: clients/horizonclient/claimable_balance_request.go [hunk 3]
```diff
@@ -16,19 +17,24 @@ func (cbr ClaimableBalanceRequest) BuildURL() (endpoint string, err error) {
...
 		queryParams := addQueryParams(
-			map[string]string{
-				"claimant": cbr.Claimant,
-				"sponsor":  cbr.Sponsor,
-				"asset":    cbr.Asset,
-			},
+			params,
 		)
```
Comment (Shaptic) [inline clients/horizonclient/claimable_balance_request.go:38]: Let's also keep the query-specific filters (claimant, etc.) separate from the "built-in" filters (cursor, limit). Something like:

```go
params := map[string]string{
    "claimant": cbr.Claimant,
    "sponsor":  cbr.Sponsor,
    "asset":    cbr.Asset,
}

queryParams := addQueryParams(params, cursor(cbr.Cursor), limit(cbr.Limit))
```

File: clients/horizonclient/claimable_balance_request_test.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+	cbr = ClaimableBalanceRequest{
+		ID:       "1235",
+		Claimant: "CLAIMANTADDRESS",
+	}
+	_, err = cbr.BuildURL()
+	assert.EqualError(t, err, "invalid request: too many parameters")
+
+	//if you have two parameters, and neither of them are ID, it must use both in the URL
+	cbr = ClaimableBalanceRequest{
+		Claimant: "CLAIMENTADDRESS",
```
Comment (Shaptic) [inline clients/horizonclient/claimable_balance_request_test.go:28]: nitpick on a typo: claim**e**nt vs. claim**a**nt

**Review by Shaptic — COMMENTED** [review]
Thank you for your pull request! I've added a few small comments below, mostly to align the implementation here with the way the other request builders work.

File: clients/horizonclient/claimable_balance_request.go (hunk 1 above)
Comment (iateadonut) [reply clients/horizonclient/claimable_balance_request.go:20]: Yes it is necessary; you can have more than one parameter if the ID is not set.

File: clients/horizonclient/claimable_balance_request.go (hunk 2 above)
Comment (iateadonut) [reply clients/horizonclient/claimable_balance_request.go:35]: Horizon still handles the defaults.  If the user gives a limit, then that limit is used.  The default is 10, but the user can set it up to 200.

This is a problem other users have had; here the user works around by doing an http request:
https://github.com/altugbakan/stellar-quest-go/blob/main/series4/quest6/solution.go#L53

File: clients/horizonclient/claimable_balance_request.go (hunk 3 above)
Comment (iateadonut) [reply clients/horizonclient/claimable_balance_request.go]: done.

File: clients/horizonclient/claimable_balance_request.go (hunk 2 above)
Comment (iateadonut) [reply clients/horizonclient/claimable_balance_request.go]: you may be referring to the second parameter of strconv.FormatUint(), which indicates the base of the number, not the default.
anyway, it's a moot point as we got rid of this.

Comment (Shaptic) [reply clients/horizonclient/claimable_balance_request.go]: Oh wow yep, completely misread this line of code: saw it as "if cbr.Limit <= 0" :facepalm:

File: clients/horizonclient/claimable_balance_request.go (hunk 1 above)
Comment (Shaptic) [reply clients/horizonclient/claimable_balance_request.go:24]: Huh, so the comment is a lie then :thinking: can you rephrase it? Then :heavy_check_mark: from my perspective :+1:

**Review by Shaptic — APPROVED** [review]
Thank you again for the PR!

File: clients/horizonclient/claimable_balance_request.go (hunk 1 above)
Comment (iateadonut) [reply clients/horizonclient/claimable_balance_request.go:24]: updated the comment to:
// You can't mix an ID query and filters.

and the error verbiage to:
return endpoint, errors.New("invalid request: you cannot have an ID query with filters")
